paper/scripts/run_counterfactuals.py

```diff
@@ -0,0 +1,383 @@
+                # use memory as input and get predictions
+                aux_memory, _ = next(iter(mem_loader))
+                aux_memory = aux_memory.cuda()
+                memory_logits = model(memory_input,aux_memory)
+                memory_predictions =  torch.argmax(memory_logits, axis=1)
+
+
+            ####################################################
+            ########## Memory Counterfactuals ##################
+            ####################################################
+            # find counterfactuals candidates among the memory samples
+            sorted_weights, sorted_weights_indices = torch.sort(weights,descending=True)
+            counterfactuals_candidates = get_counterfactuals_indices(sorted_weights_indices,
+                        memory_predictions, predicted_class)
+
+            if len(counterfactuals_candidates) == 0:
+                # rare case of no candidates in memory
+                continue
+
+            # count how many candidates have a weight greater then zero
+            count_counterfactuals = torch.count_nonzero(counterfactuals_candidates*sorted_weights,1)
+
+            index_samples_with_cf = count_counterfactuals.nonzero()
+            samples_with_cf = (counterfactuals_candidates*sorted_weights)[index_samples_with_cf]
+
+
+            if samples_with_cf.size()[0] > 1:
+
+                # extract the counterfactual associated with the highest weight
+                index_highest_weight= torch.argmax(samples_with_cf,2)
+                index_top_counterfactual = sorted_weights_indices[0,index_highest_weight]
+
+                prediction_top_counterfactual = memory_predictions[index_top_counterfactual]
+
+
+                target_class = prediction_top_counterfactual
+
+
+
+                if FLAGS.algo == 'proto':
+                    with torch.no_grad():
+                        prototype = get_closest_prototypes(
+                            ae_fulldataset.encoder.cpu(),
+                            img_gpu[torch.flatten(index_samples_with_cf)].cpu(),
+                            class_prototypes, None, k, k_type, target_class)
+                        prototype = prototype.cuda()
+                    ae_fulldataset =  ae_fulldataset.cuda()
+                    counterfactual = counterfactuals_utils.attack(model,
+                            img_gpu[torch.flatten(index_samples_with_cf)],
+                            onehot_prediction[torch.flatten(index_samples_with_cf)],
+                            memory=memory_input,
+                            enc_model=ae_fulldataset.encoder,
+                            target_class=target_class, class_proto=class_prototypes,
+                            feature_range=feature_range, proto_val=prototype,
+                            verbose=False)
+                else :
+                    # case algo = 'memory'
+                    counterfactual = memory_input[index_top_counterfactual]
+
+
```

Comment:
all this emtpy space here is confusing

Reply:
Fixed in [81c3488](https://github.com/KRLGroup/memory-wrap/pull/3/commits/81c3488bd3ffcc613a4d887bfefda7dadcf37dc8)